### PR TITLE
Enh gen tractogram and sampling methods

### PR DESCRIPTION
**Pull request by ppoulin91**

### CODE REVIEW

File: learn2track/models/gru.py [hunk 1]
```diff
@@ -143,10 +152,19 @@ def load(self, path):
...
-        model = cls(**hyperparams)
-        model.load(path)
+        try:
+            loaddir = pjoin(path, cls.__name__)
+            hyperparams = smartutils.load_dict_from_json_file(pjoin(loaddir, "hyperparams.json"))
+            hyperparams.update(kwargs)
+
+            model = cls(**hyperparams)
+            model.load(path)
+        except FileNotFoundError:
```
Comment (ppoulin91) [inline learn2track/models/gru.py]: What is your opinion on this? I often had to evaluate/generate using an unfinished model, but I am not sure if this is the right way, or if it at least should be verbose when this is the case

Comment (MarcCote) [reply learn2track/models/gru.py]: Usually, instead of pointing to `experiments/my_experiment/`, I point directly to `experiments/my_experiment/training`

Comment (MarcCote) [reply learn2track/models/gru.py]: That means the burden is put on every script that uses experiments. If your approach is working correctly, that would be pretty neat and help cleaning the scripts. Just add a print saying the unfinished model is being used.

But now that I thought about it, when is this happening? That is, you have a model save in training/ but not in the root folder of the experiment?

File: learn2track/models/gru_mixture.py [hunk 2]
```diff
@@ -77,6 +77,15 @@ def _get_mixture_samples(self, srng, mixture_weights, means, stds):
...
+    def _get_max_component_samples(self, mixture_weights, means, stds):
+        batch_size = mixture_weights.shape[0]
+        xs = T.arange(0, batch_size)
+
+        choices = T.argmax(mixture_weights, axis=1)
+        samples = means[xs, choices]
+
+        return samples
+
     def seq_next(self, x_t, subject_ids=None):
```
Comment (MarcCote) [inline learn2track/models/gru_mixture.py:89]: While at it, should we remove the `seq_next`, `seq_reset` methods? Or is it better to do it in another PR?

File: learn2track/models/gru_mixture.py [hunk 3]
```diff
@@ -276,14 +280,65 @@ def _compute_losses(self, model_output):
         # means.shape : (batch_size, seq_len, n_gaussians, 3)
 
         # expected_value.shape : (batch_size, seq_len, 3)
```
Comment (MarcCote) [inline learn2track/models/gru_mixture.py]: expected_value => samples

File: learn2track/models/gru_mixture.py [hunk 4]
```diff
@@ -50,15 +50,15 @@ def hyperparameters(self):
         hyperparameters['layer_regression_size'] = self.layer_regression_size
         return hyperparameters
 
-    def get_mixture_parameters(self, regression_output):
+    def _get_mixture_parameters(self, regression_output):
```
Comment (MarcCote) [inline learn2track/models/gru_mixture.py]: Let's make it more generic:
```python
def _get_mixture_parameters(self, regression_output):
    shape = regression_output.shape[:-1]
    mixture_weights = T.nnet.softmax(regression_output[..., :self.n_gaussians])
    means = T.reshape(regression_output[..., self.n_gaussians:self.n_gaussians * 4], shape + (self.n_gaussians, 3))
    stds = T.reshape(T.exp(regression_output[..., self.n_gaussians * 4:self.n_gaussians * 7]), shape + (self.n_gaussians, 3))

    return mixture_weights, means, stds
```

File: learn2track/models/gru_mixture.py [hunk 5]
```diff
@@ -276,14 +280,65 @@ def _compute_losses(self, model_output):
...
+        self.samples = T.sum(mixture_weights[:, :, :, None] * means, axis=2)
+
+        # loss_per_time_step.shape = (batch_size, seq_len)
+        self.loss_per_time_step = T.sqrt(T.sum(((self.samples - self.dataset.symb_targets)**2), axis=2))
+        # loss_per_seq.shape = (batch_size,)
+        self.loss_per_seq = T.sum(self.loss_per_time_step*mask, axis=1) / T.sum(mask, axis=1)
+
+        return self.loss_per_seq
+
+    def _get_mixture_parameters(self, regression_output):
```
Comment (MarcCote) [inline learn2track/models/gru_mixture.py]: Not needed. We should use the model's method (see related comment about a generic version of that method in the GRUMixture class).

File: learn2track/models/gru_mixture.py [hunk 6]
```diff
@@ -276,14 +280,65 @@ def _compute_losses(self, model_output):
...
+        self.samples = means[row_ids, column_ids, choices]
+
+        # loss_per_time_step.shape = (batch_size, seq_len)
+        self.loss_per_time_step = T.sqrt(T.sum(((self.samples - self.dataset.symb_targets)**2), axis=2))
+        # loss_per_seq.shape = (batch_size,)
+        self.loss_per_seq = T.sum(self.loss_per_time_step*mask, axis=1) / T.sum(mask, axis=1)
+
+        return self.loss_per_seq
 
     def _get_mixture_parameters(self, regression_output):
```
Comment (MarcCote) [inline learn2track/models/gru_mixture.py]: Not needed. We should use the model's method (see related comment about a generic version of that method in the GRUMixture class).

File: learn2track/models/gru_mixture.py [hunk 7]
```diff
@@ -276,14 +280,65 @@ def _compute_losses(self, model_output):
...
+        regression_outputs = model_output
+
+        # mixture_weights.shape : (batch_size, seq_len, n_gaussians)
+        # means.shape : (batch_size, seq_len, n_gaussians, 3)
+        mixture_weights, means, stds = self._get_mixture_parameters(regression_outputs)
+
+        batch_size = mixture_weights.shape[0]
+        seq_len = mixture_weights.shape[1]
+
+        row_ids = T.tile(T.arange(0, batch_size), (seq_len,1)).T    # [[0,0,...,0],[1,1,...,1],...]
```
Comment (MarcCote) [inline learn2track/models/gru_mixture.py]: Neat. But I **think** you don't need the `T.tile` and just rely on broadcasting. I might be wrong, though. Something, along the lines of
```python
    # Compute NLLs
    # predictions.shape: (batch_size, seq_length, out_size)
    neg_log_probs = -T.log(predictions)
    nlls = neg_log_probs[T.arange(class_ids.shape[0])[:, None],
                         T.arange(class_ids.shape[1])[None, :],
                         class_ids[:, :, 0]]
```

Comment (MarcCote) [reply learn2track/models/gru_mixture.py]: Not mandatory.

File: learn2track/models/gru_msp.py [hunk 8]
```diff
@@ -128,9 +128,12 @@ def _get_sample(distribution_parameters, noise):
...
         sigma = distribution_parameters[:, 3:6]
 
-        sample = mu + noise * sigma
+        samples = mu + noise * sigma
 
-        return sample
+        return samples
+
+    def _get_max_component_samples(self, distribution_parameters):
+        return distribution_parameters[:, :3]
```
Comment (MarcCote) [inline learn2track/models/gru_msp.py]: Are those values represent the means? What is the shape of `distribution_parameters` ?

File: learn2track/models/gru_regression.py [hunk 9]
```diff
@@ -130,15 +129,7 @@ def use(self, X):
...
-    def get_init_states(self, batch_size):
-        states_h = []
-        for i, hidden_size in enumerate(self.hidden_sizes):
-            state_h = np.zeros((batch_size, hidden_size), dtype=floatX)
-            states_h.append(state_h)
-
-        return states_h
-
-    def make_sequence_generator(self, subject_id=0):
+    def make_sequence_generator(self, subject_id=0, **kwargs):
```
Comment (MarcCote) [inline learn2track/models/gru_regression.py]: I prefer using `**_` instead of `**args` to make it clear the additional arguments aren't going to be used.

File: learn2track/models/gru_regression.py [hunk 10]
```diff
@@ -219,9 +210,11 @@ def _compute_losses(self, model_output):
...
             regression_outputs /= (T.sqrt(T.sum(regression_outputs**2, axis=2, keepdims=True) + self.eps))
 
-        # L2_errors_per_time_step.shape = (batch_size,)
-        self.L2_errors_per_time_step = T.sqrt(T.sum(((regression_outputs - self.dataset.symb_targets)**2), axis=2))
-        # avg_L2_error_per_seq.shape = (batch_size,)
-        self.avg_L2_error_per_seq = T.sum(self.L2_errors_per_time_step*mask, axis=1) / T.sum(mask, axis=1)
+        self.samples = regression_outputs
+
+        # loss_per_time_step.shape = (batch_size,)
+        self.loss_per_time_step = T.sqrt(T.sum(((self.samples - self.dataset.symb_targets)**2), axis=2))
```
Comment (MarcCote) [inline learn2track/models/gru_regression.py]: This line and the one below are being used a lot (different loss implementation for different models). It can wait, but it would be easy to refactored it out in a `losses_utils.py`.

File: learn2track/models/gru.py (hunk 1 above)
Comment (ppoulin91) [reply learn2track/models/gru.py]: Exactly, since the model is saved in `experiments/my_experiment/ ` only once all training is done. Given a batch of experiments which take a long time to run (e.g. more than 24 hours on helios), you might want to evaluate all of them even if some haven't finished.

File: learn2track/models/gru_mixture.py (hunk 2 above)
Comment (ppoulin91) [reply learn2track/models/gru_mixture.py:89]: I'd rather do the cleanup in another PR

**Review by MarcCote — COMMENTED** [review]
A few comments but overall very nice. Good job!

File: learn2track/models/gru.py (hunk 1 above)
Comment (MarcCote) [reply learn2track/models/gru.py]: I thought I'd changed it!

Comment (MarcCote) [reply learn2track/models/gru.py]: It seems it was for another project! Here what we should have in `learn.py`:

```python

        # Callback function to stop training if NaN is detected.
        def detect_nan(obj, status):
            if np.isnan(model.parameters[0].get_value().sum()):
                print("NaN detected! Stopping training now.")
                sys.exit()
        trainer.append_task(tasks.Callback(detect_nan, each_k_update=1))

        # Callback function to save training progression.
        def save_training(obj, status):
            trainer.save(experiment_path)

        trainer.append_task(tasks.Callback(save_training))

        # Early stopping with a callback for saving every time model improves.
        def save_improvement(obj, status):
            """ Save best model and training progression. """
            if np.isnan(model.parameters[0].get_value().sum()):
                print("NaN detected! Not saving the model. Crashing now.")
                sys.exit()

            print("*** Best epoch: {0} ***\n".format(obj.best_epoch))
            model.save(experiment_path)

```

Comment (MarcCote) [reply learn2track/models/gru.py]: I'll add it, once this PR is merged.

Comment (ppoulin91) [reply learn2track/models/gru.py]: Discussion result : Let's remove it!

File: learn2track/models/gru_mixture.py (hunk 4 above)
Comment (ppoulin91) [reply learn2track/models/gru_mixture.py]: Check array.ndim if ellipsis doesn't work

**Comment by MarcCote** [comment]
Nice job!
